Assets/MixedRealityToolkit.SDK/Experimental/Features/UX/MixedRealityKeyboardBase.cs

```diff
@@ -0,0 +1,324 @@
+            }, false);
+#endif
+        }
+
+#if WINDOWS_UWP
+        private IEnumerator UpdateState()
+        {
+            while (true)
+            {
+                switch (state)
+                {
+                    case KeyboardState.Showing:
+                        {
+                            UpdateText();
+                        }
+                        break;
+
+                    case KeyboardState.Hiding:
+                        {
+                            if (onHideKeyboard != null)
+                            {
+                                onHideKeyboard.Invoke();
+                            }
+                        }
+                        break;
+                }
+
+                yield return null;
+            }
+        }
+#endif
+
+        private void OnDisable()
+        {
+            HideKeyboard();
+        }
+
+#endregion MonoBehaviour Implementation
+
+        /// <summary>
+        /// Closes the keyboard for user interaction.
+        /// </summary>
+        public void HideKeyboard()
+        {
+            if (state != KeyboardState.Hidden)
+            {
+                state = KeyboardState.Hidden;
+            }
+
+#if WINDOWS_UWP
+            UnityEngine.WSA.Application.InvokeOnUIThread(() => inputPane?.TryHide(), false);
+
+            if (stateUpdate != null)
+            {
+                StopCoroutine(stateUpdate);
+                stateUpdate = null;
+            }
+#endif
+        }
+
+        /// <summary>
+        /// Opens the keyboard for user interaction.
+        /// </summary>
+        /// <param name="text">Initial text to populate the keyboard with.</param>
+        /// <param name="multiLine">True, if the return key should signal a newline rather than a commit.</param>
+        public virtual void ShowKeyboard(string text = "", bool multiLine = false)
+        {
+            Text = text;
+            this.multiLine = multiLine;
+
+            // 2019/08/14: We show the keyboard even when the keyboard is already visible because on HoloLens 1
+            // and WMR the events OnKeyboardShowing and OnKeyboardHiding do not fire
+            //if (state == KeyboardState.Showing)
+            //{
+            //    Debug.Log($"MixedRealityKeyboard.ShowKeyboard called but keyboard already visible.");
+            //    return;
+            //}
+
+            state = KeyboardState.Showing;
+
+#if WINDOWS_UWP
+            if (keyboard != null)
+            {
+                keyboard.text = Text;
+                UnityEngine.WSA.Application.InvokeOnUIThread(() => inputPane?.TryShow(), false);
+            }
+            else
+            {
+                keyboard = TouchScreenKeyboard.Open(Text, TouchScreenKeyboardType.Default, false, this.multiLine, false, false);
+            }
+
+            if (onShowKeyboard != null)
+            {
+                onShowKeyboard.Invoke();
+            }
```

Comment:
Again don't block on getting PR but FYI for future
Simplify to 

onShowKeyboard?.Invoke();